### ISSUE TITLE
multi-sig transaction creation

### ISSUE THREAD

**Issue by jeffallen** [dedis/cothority#1758]
bcadmin (and other clients) needs to support the creation of multi-sig transactions, so that we can implement co-signed transactions that can pass DARCs with AND in them.

We need to think out how this would work from a UI point of view, and if we can use the ledger itself as a rendezvous point to make this easier to do.

At the same time, perhaps this rendezvous point could contribute to solving the problem wherein if I use my signercount 5 for a signature someone else will contribute to, I cannot send any other txns until we send the co-signed txn, because if I send a new signercount of 5, I invalidate my signature on the co-signed txn, and if I send a txn signed with signercount 6 before the co-signed txn is ready and posted, then it will be refused. Perhaps I could post a commit saying, "you haven't seen it yet, but I promise that a txn with signercount X is coming, and my signature will match this hash, so please let me start using signercount x+1 already".

**Comment by ineiti** [comment]
Yes, good point, which I brought up by email for the recovery method I'm implementing. Some ideas for continuing the discussion:

## Allow -1 for signercount

At least in my example, the darc-rule to be executed would be unique anyway and replay-protected by the fact that it updates a darc, and that you cannot downgrade a version of a darc. So we could allow a -1 for signercount (or MAX_UINT64) to note that the smart contract is taking care of the replay-protection.

## Ephemeral Keys

One solution is to create ephemeral keypairs for signatures you're not expecting to happen for some time. In my example, where a user wants to give recovery access to a set of contacts, he could do the following:

Darc_user::recover = darc_trustee1 AND darc_trustee2

And when trustee1 and trustee2 want to sign, they add an ephemeral keypair to their darc:

Darc_trustee1::sign = public_longterm1 OR public_ephemeral1
Darc_trustee2::sign = public_longterm2 OR public_ephemeral2

Now they can sign using this ephemeral key, and can still continue using their public keys. Of course this is not ideal.

## Store multiple signercounts

Instead of storing the latest signercount and only allowing the next signercount, store something like:

- 0..10, 12..15, 17..20

So that you can allow signercounts that are not the previous count + 1.

**Comment by jeffallen** [comment]
@Gilthoniel had an interesting idea. When a txn is refused, it ends up on the chain anyway. So what if the first signer sends in the txn, knowing it will fail. He gets back a block # and a short txn hash, like (23-7a7f7e7d73). He shares that with the other admins, and those who want use "bcadmin txn sign 23-7a7f7e7d73", which fetches the txn, displays it, asks if you really want to sign, and if so adds the signature and resubmits it. If the DARC is looking for 2 signers, then it will be accepted. If it is refused, you get a new #/hash pair, and can pass that to the 3rd person who should sign.

Summary: (ab)use the transaction refused log as a RDV point.

**Comment by ineiti** [comment]
I like it, looks backward-compatible - some things to take into account:

1. for n signers you need to wait n blocks
2. you cannot collect the signatures offline (well, you could use my `ephemeral keys` for that, which doesn't require you to change anything)
3. you need to know the signers in advance, so if you make a 5 out of 7 signature, you need to know in advance which 5 will participate, because the public keys will show up in the hash that will be signed. Even though you could add all possible keys to the hash, but that might be too much, if you have many DARCs linked to each other.

So it's mostly 3. I think we need to think about, which is in fact the use-case I'm thinking of for the recovery mechanism I'm currently writing.

**Comment by Gilthoniel** [comment]
3. Do you ? Isn't it up to the DARC ? Like if you have *A* OR *B* you need one signature and if you have *A* AND *B* you will need two ? If you have multiple AND, the transaction will be accepted with the minimum requirement and then remaining signatures are not necessary anyway (if there are the darc is wrong)

**Comment by jeffallen** [comment]
Oooh, big problem. As @ineiti mentioned, you have to have all the SignerIdentities that could be needed to sign present already in the instruction before anyone signs it.

But look at what else is in the hash:
```
// Hash computes the digest of the hash function
func (instr Instruction) Hash() []byte {
	h := sha256.New()
	h.Write(instr.InstanceID[:])
	var args []Argument
	switch instr.GetType() {
	case SpawnType:
		h.Write([]byte{0})
		h.Write([]byte(instr.Spawn.ContractID))
		args = instr.Spawn.Args
	case InvokeType:
		h.Write([]byte{1})
		h.Write([]byte(instr.Invoke.ContractID))
		args = instr.Invoke.Args
	case DeleteType:
		h.Write([]byte{2})
		h.Write([]byte(instr.Delete.ContractID))
	}
	for _, a := range args {
		nameBuf := []byte(a.Name)
		nameLenBuf := make([]byte, 8)
		binary.LittleEndian.PutUint64(nameLenBuf, uint64(len(nameBuf)))
		h.Write(nameLenBuf)
		h.Write(nameBuf)

		valueLenBuf := make([]byte, 8)
		binary.LittleEndian.PutUint64(valueLenBuf, uint64(len(a.Value)))
		h.Write(valueLenBuf)
		h.Write(a.Value)
	}
	for _, ctr := range instr.SignerCounter {
		ctrBuf := make([]byte, 8)
		binary.LittleEndian.PutUint64(ctrBuf, ctr)
		h.Write(ctrBuf)
	}
	for _, id := range instr.SignerIdentities {
		buf := id.GetPublicBytes()
		lenBuf := make([]byte, 8)
		binary.LittleEndian.PutUint64(lenBuf, uint64(len(buf)))
		h.Write(lenBuf)
		h.Write(buf)
	}
	return h.Sum(nil)
}
```

You need to know *which signer counter* each signer *will be using* before any of them can sign it. 

You'd need a two-phase system, where anyone who is going to sign it says, "I'll be signing with counter C", then compile the list of signers and signer counters, then collects the signatures.

**Comment by ineiti** [comment]
Yes that works for 1 out of 2, but there we don't have a problem in the first place ;)

Let's say you have a 2 out of 3, that's:
`(A AND B) OR (A AND C) OR (B AND C)`
So if you create the signature, you will sign the hash, which includes the Signers. And if you don't know beforehand which of the `OR`s you're handling, you have to put A, B, and C in the Signers. 

But in fact, you have to rewrite the signature-verification in ByzCoin anyway, so all you need to do is to throw away signers in the instruction that don't have a valid signature, merge it with the signers from previous, failed attempts, and then only pass these to the darc-verification in here:

https://github.com/dedis/cothority/blob/master/byzcoin/transaction.go#L262

So, yes, I think that should work.

How would you pass the `{blockNbr, tx-hash}` in the instruction? Could it be just a special `Signature`, as this is only a slice of bytes anyway? Then you wouldn't need to change too much, and it would be an extension, not a change.

**Comment by nkcr** [comment]
Using this space to document the current design process. I will update it along the way.

**Terminology**:

- **Client**: someone that interacts with ByzCoin by sending transactions
- **Root instruction**: the instruction we want to execute that require more than one client to sign

**Note**:

- For simplification purpose we consider that each transaction holds only one instruction. Hence, we are most of the time using the term *instruction* where it would be more appropriate to use the term *transaction*. 

We've been considering 4 designs:

# Design 1 "multisign links"

**Main idea**: The identities of the clients that agree to sign a root instruction are stored in the form of a linked list.

**Example**: If the root instruction must satisfies `A & B & C`, then:
- `A` sends the normal root instruction with its own signature. The instruction will fail since it requires `A & B & C`, but it will stay on the skip chain. 
- `A` asks `B` to send a `multisign` instruction that refers to the failed one that `A` previously sent. The instruction that `B` sends is of form `multisign:block_id:instruction_id`. This instruction says "*hey, I am B and I agree on the root instruction stored at the `block_id:instruction_id` space*".
- `A` asks `C` to send a `multisign` instruction that refers to the one that `B` sent. This instruction says "*Hey, I am C and I agree on the root instruction that is stored at the begining of the link*".

Each time a `multisign` instruction is sent, we can retrieve the list of signers that agreed on the root instruction. Once the list of signers satisfies the root instruction, we can execute it.

**White board**:

![IMG_20190424_130129](https://user-images.githubusercontent.com/1918021/56719281-b98b6900-6740-11e9-90fb-ddd101faeb38.jpg)

**Challenges**:

- Since this is "stateless", how can we say that the root transaction has already been executed? For example, if `C` sends its `multisign` instruction twice, it will then execute the root instruction twice.
- This is strange that a `multisign` instruction could actually trigger a completely non-related state change by executing the root instruction.

# Design 2 "Rendez-vous contract"

**Main idea**: Implement a new contract that acts as a rendez-vous point for the clients.

**Concept**: Lets detail the `Spawn`, `Invoke`, and `Delete` method of this contract:

- `Spawn`: Take as input the root instruction
- `Invoke`: Actually has two methods:
  - `Invoke:addProof`: Allows a client to sign the root instruction. The client's proof is added to a current list of proofs.
  - `Invoke:execRoot`: Based on the current list of proofs, executes the root instruction.
- `Delete`: Removes the content of the instance. This is used when the root instruction has been executed to clear the mess and prevent a replay.

**Challenges**:

- Handle instructions atomically to avoid collision if two clients add a proof at the same time, or if two client invoke the `execRoot` instruction simultaneously. 
- Override correctly the `VerifyInstruction` method of this contract.
- We can not use the clients counters when they sign the root instruction since they are likely to be different at the time the root instruction will be executed. It may be OK not to take them into account "inside the contract" to sign the root instruction since the counters are used when the clients sends instruction to the contract (hence, all that is coming from the outside is verified). This implies that each type of root instruction must implement a `verifyInstruction` that do not take counters into account for this special case (ie. coming from our special contract). 

# Design 3 "Key-value agreement"

**Main idea**: Each client use the same key-value contract to agree on a root command and tell everyone which counter will be used. Once all the counters are known, each client can generate a signature based on the root instruction and the counters.

**Challenges**: 

- Each client must promise a counter value. Which means that it can not perform any other transactions in-between
- If a client wants to change its promised counter, everyone must re-calculate its signature
- ...

# Design 4 "Ephemeral keys"

To be continued...
